src/botservice/azext_bot/custom.py

```diff
@@ -0,0 +1,466 @@
+        else:
+            zip_url = 'https://connectorprod.blob.core.windows.net/bot-packages/node.js-abs-functions_emptybot_funcpack.zip'
+
+    else:
+        kind = 'sdk'
+        template_name = 'webapp.template.json'
+        if language == 'Csharp':
+            zip_url = 'https://connectorprod.blob.core.windows.net/bot-packages/csharp-abs-webapp_simpleechobot_precompiled.zip'
+        else:
+            zip_url = 'https://connectorprod.blob.core.windows.net/bot-packages/node.js-abs-webapp_hello-chatconnector.zip'
+
+    create_new_storage = False
+    if not storageAccountName:
+        create_new_storage = True
+        import re
+        import string
+        import random
+        storageAccountName = re.sub(r'[^a-z0-9]','',resource_name[:10] + ''.join(random.choice(string.ascii_lowercase+string.digits) for _ in range(4)))
+
+    paramsdict = {
+    "location" : location,
+    "kind" : kind,
+    "sku" : sku,
+    "siteName" : resource_name,
+    "description" : description,
+    "appId" : appid,
+    "appSecret" : password,
+    "storageAccountResourceId" : "",
+    "serverFarmId" : "/subscriptions/{0}/resourceGroups/{1}/providers/Microsoft.Web/serverfarms/{2}".format(client.config.subscription_id, resource_group_name, resource_name),
+    "zipUrl" : zip_url,
+    "createNewStorage" : create_new_storage,
+    "storageAccountName" : storageAccountName,
+    "botEnv" : "prod",
+    "useAppInsights" : True,
+    "appInsightsLocation" : appInsightsLocation,
+    "createServerFarm" : True,
+    "serverFarmLocation": location.lower().replace(' ', ''),
+    "azureWebJobsBotFrameworkDirectLineSecret" : "",
+    "botId" : resource_name
+    }
+    params = {k: {'value' : v} for k,v in paramsdict.items()}
+
+    import os
+    import json
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    deploy_result = deploy_arm_template(
+        cmd = cmd,
+        resource_group_name = resource_group_name,
+        template_file = '{0}\\{1}'.format(dir_path, template_name),
+        parameters = [[json.dumps(params)]],
+        deployment_name = resource_name,
+        mode = 'Incremental'
+    )
+
+    deploy_result.wait()
+    if bot_json:
+        return create_bot_json(cmd, client, resource_group_name, resource_name, app_password=password)
+
+def publish_app(cmd, client, resource_group_name, resource_name, giturl = None, git_token = None, git_branch = 'master', code_dir = None):
+    #if given msbot json, use that to update environment settings like luis settings
+    if giturl:
+        return config_source_control(
+            cmd = cmd,
+            name = resource_name,
+            resource_group_name = resource_group_name,
+            repo_url = giturl,
+            branch = git_branch,
+            git_token = git_token
+        )
+
+    #since there is no git url it's definitely publish from local
+    if not code_dir:
+        import os
```

Comment:
You can really just import os at the very top of the file instead of importing it so many times.
Same with other standard python imports such as `shutil`

Reply:
good catch! thanks. .updated the code to move os, shutil and json imports to the top of the file